File: skirk/transports/sheets.py
"""Sheets-backed low-rate BlobStore.

This treats a spreadsheet range as an append-only log. Deletes are tombstones,
not physical row removal, because the simple Values API is much safer than
structural sheet mutation for this prototype.
"""

from __future__ import annotations

import base64
import json
import time
from urllib.parse import quote, urlencode

from skirk.curl import CurlHttpClient, GoogleRoute, gcloud_access_token

from .base import ObjectInfo, TransportError


class SheetsRingStore:
    host = "sheets.googleapis.com"

    def __init__(
        self,
        *,
        spreadsheet_id: str,
        range_name: str = "skirk!A:D",
        proxy: str = "socks5h://127.0.0.1:1080",
        google_ip: str = "216.239.38.120",
        route_mode: str = "real-pinned",
        token: str | None = None,
    ):
        self.spreadsheet_id = spreadsheet_id
        self.range_name = range_name
        self.token = token or gcloud_access_token()
        self.client = CurlHttpClient(GoogleRoute(proxy=proxy, google_ip=google_ip, mode=route_mode))

    def _auth(self) -> dict[str, str]:
        return {"Authorization": f"Bearer {self.token}"}

    def _append_row(self, row: list[str]) -> None:
        encoded_range = quote(self.range_name, safe="")
        params = urlencode({"valueInputOption": "RAW", "insertDataOption": "INSERT_ROWS"})
        body = json.dumps({"majorDimension": "ROWS", "values": [row]}).encode("utf-8")
        result = self.client.request(
            method="POST",
            host=self.host,
            path=f"/v4/spreadsheets/{self.spreadsheet_id}/values/{encoded_range}:append?{params}",
            headers=self._auth() | {"Content-Type": "application/json"},
            body=body,
        )
        if not result.ok:
            raise TransportError(f"Sheets append failed status={result.status}: {result.body[:500]!r}")

    def _rows(self) -> list[list[str]]:
        encoded_range = quote(self.range_name, safe="")
        result = self.client.request(
            method="GET",
            host=self.host,
            path=f"/v4/spreadsheets/{self.spreadsheet_id}/values/{encoded_range}",
            headers=self._auth(),
        )
        if not result.ok:
            raise TransportError(f"Sheets read failed status={result.status}: {result.body[:500]!r}")
        payload = json.loads(result.body.decode("utf-8"))
        return payload.get("values", [])

    def put(self, name: str, data: bytes) -> None:
        encoded = base64.urlsafe_b64encode(data).decode("ascii")
        self._append_row([name, encoded, str(time.time_ns()), "put"])

    def get(self, name: str) -> bytes:
        latest: list[str] | None = None
        for row in self._rows():
            if len(row) >= 4 and row[0] == name:
                latest = row
        if not latest or latest[3] == "delete":
            raise FileNotFoundError(name)
        return base64.urlsafe_b64decode(latest[1].encode("ascii"))

    def list(self, prefix: str) -> list[ObjectInfo]:
        latest: dict[str, list[str]] = {}
        for row in self._rows():
            if len(row) >= 4 and row[0].startswith(prefix):
                latest[row[0]] = row
        return [
            ObjectInfo(name=name, size=len(row[1]), updated=row[2])
            for name, row in sorted(latest.items())
            if row[3] != "delete"
        ]

    def delete(self, name: str) -> None:
        self._append_row([name, "", str(time.time_ns()), "delete"])

File: internal/skirk/sheets.go
package skirk

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"
)

type SheetsLog struct {
	http          *GoogleHTTPClient
	token         string
	spreadsheetID string
	rangeName     string
}

func NewSheetsLog(httpClient *GoogleHTTPClient, token string, cfg SheetsConfig) *SheetsLog {
	rangeName := cfg.Range
	if rangeName == "" {
		rangeName = "skirk!A:D"
	}
	return &SheetsLog{http: httpClient, token: token, spreadsheetID: cfg.SpreadsheetID, rangeName: rangeName}
}

func (s *SheetsLog) Put(ctx context.Context, name string, data []byte) error {
	return s.append(ctx, []string{name, base64.URLEncoding.EncodeToString(data), strconv.FormatInt(time.Now().UnixNano(), 10), "put"})
}

func (s *SheetsLog) Get(ctx context.Context, name string) ([]byte, error) {
	rows, err := s.rows(ctx)
	if err != nil {
		return nil, err
	}
	var latest []string
	for _, row := range rows {
		if len(row) >= 4 && row[0] == name {
			latest = row
		}
	}
	if latest == nil || latest[3] == "delete" {
		return nil, fmt.Errorf("sheet row not found: %s", name)
	}
	return base64.URLEncoding.DecodeString(latest[1])
}

func (s *SheetsLog) List(ctx context.Context, prefix string) ([]ObjectInfo, error) {
	rows, err := s.rows(ctx)
	if err != nil {
		return nil, err
	}
	latest := map[string][]string{}
	for _, row := range rows {
		if len(row) >= 4 && strings.HasPrefix(row[0], prefix) {
			latest[row[0]] = row
		}
	}
	var infos []ObjectInfo
	for name, row := range latest {
		if row[3] == "delete" {
			continue
		}
		infos = append(infos, ObjectInfo{Name: name, Size: int64(len(row[1])), Updated: row[2]})
	}
	sort.Slice(infos, func(i, j int) bool { return infos[i].Name < infos[j].Name })
	return infos, nil
}

func (s *SheetsLog) Delete(ctx context.Context, name string) error {
	return s.append(ctx, []string{name, "", strconv.FormatInt(time.Now().UnixNano(), 10), "delete"})
}

func (s *SheetsLog) append(ctx context.Context, row []string) error {
	encodedRange := url.PathEscape(s.rangeName)
	values := url.Values{}
	values.Set("valueInputOption", "RAW")
	values.Set("insertDataOption", "INSERT_ROWS")
	body, err := json.Marshal(map[string]any{
		"majorDimension": "ROWS",
		"values":         [][]string{row},
	})
	if err != nil {
		return err
	}
	path := fmt.Sprintf("/v4/spreadsheets/%s/values/%s:append?%s", url.PathEscape(s.spreadsheetID), encodedRange, values.Encode())
	result, err := s.http.Request(ctx, http.MethodPost, "sheets.googleapis.com", path, s.jsonHeaders(), body)
	if err != nil {
		return err
	}
	return require2xx(result, "sheets append")
}

func (s *SheetsLog) rows(ctx context.Context) ([][]string, error) {
	encodedRange := url.PathEscape(s.rangeName)
	path := fmt.Sprintf("/v4/spreadsheets/%s/values/%s?majorDimension=ROWS", url.PathEscape(s.spreadsheetID), encodedRange)
	result, err := s.http.Request(ctx, http.MethodGet, "sheets.googleapis.com", path, s.authHeaders(), nil)
	if err != nil {
		return nil, err
	}
	if result.Status == http.StatusNotFound {
		return nil, nil
	}
	if err := require2xx(result, "sheets get"); err != nil {
		return nil, err
	}
	var payload struct {
		Values [][]string `json:"values"`
	}
	if err := json.Unmarshal(result.Body, &payload); err != nil {
		return nil, err
	}
	return payload.Values, nil
}

func (s *SheetsLog) jsonHeaders() map[string]string {
	headers := s.authHeaders()
	headers["Content-Type"] = "application/json"
	return headers
}

func (s *SheetsLog) authHeaders() map[string]string {
	return map[string]string{"Authorization": "Bearer " + s.token}
}
